StreamChatClient_v3/WebSocketClient/ConnectionState.swift

```diff
@@ -0,0 +1,62 @@
+//
+// ConnectionState.swift
+// Copyright © 2020 Stream.io Inc. All rights reserved.
+//
+
+import Foundation
+
+public typealias ConnectionId = String
+
+/// A web socket connection state.
+public enum ConnectionState: Equatable {
+  /// Provides additional information about the source of disconnecting.
+  public enum DisconnectionSource: Equatable {
+    /// A user initiated web socket disconnecting.
+    case userInitiated
+
+    /// A server initiated web socket disconnecting, an optional error object is provided.
+    case serverInitiated(error: ClientError? = nil)
+
+    /// The system initiated web socket disconnecting.
+    case systemInitiated
+  }
+
+  /// The web socket is not connected.
+  case notConnected
```

Comment:
What's the functional difference between this (`notConnected`) and `disconnected(nil)`?

Reply:
`notConnected` is the default start state, `disconnected` is the final stop state. I think we could get rid of `notConnected` and use `disconnected` as the default state. The optional error there would be a little bit weird, though.

Reply:
Optional error would signal if the connection closed due to error or intentionally, IMHO

Reply:
It was `.disconnected(nil)` by default some time ago, but it's a different state. We should keep `notConnected` for the initial state and when we are trying to reconnect: `.connecting` -> `.notConnected` (on fail).

Reply:
I'll go with `notConnected` with an optional error. I think it makes sense. @buh we have a special state `waitingForReconnect` for the situations when a reconnection attempt is scheduled.

Reply:
@VojtaStavik yes and you said you will call it directly 😆